network/identity/identity.go

```diff
@@ -0,0 +1,216 @@
+	return &IdentityService{
+		logger:     logger.Named("identity"),
+		baseServer: server,
+		chainID:    chainID,
+		hostID:     hostID,
+	}
+}
+
+func (i *IdentityService) GetNotifyBundle() *network.NotifyBundle {
+	return &network.NotifyBundle{
+		ConnectedF: func(net network.Network, conn network.Conn) {
+			peerID := conn.RemotePeer()
+			i.logger.Debug("Conn", "peer", peerID, "direction", conn.Stat().Direction)
+
+			if i.hasPendingStatus(peerID) {
+				// handshake has already started
+				return
+			}
+
+			if !i.baseServer.HasFreeConnectionSlot(conn.Stat().Direction) {
+				i.disconnectFromPeer(peerID, ErrNoAvailableSlots.Error())
+
+				return
+			}
+
+			// Mark the peer as pending (pending handshake)
+			i.addPendingStatus(peerID, conn.Stat().Direction)
+
+			go func() {
+				connectEvent := &event.PeerEvent{
+					PeerID: peerID,
+					Type:   event.PeerDialCompleted,
+				}
+
+				if err := i.handleConnected(peerID, conn.Stat().Direction); err != nil {
+					// Close the connection to the peer
+					i.disconnectFromPeer(peerID, err.Error())
+
+					connectEvent.Type = event.PeerFailedToConnect
+				}
+
+				// Mark the peer as no longer pending
+				i.removePendingStatus(connectEvent.PeerID)
+
+				// Emit an adequate event
+				i.baseServer.EmitEvent(&event.PeerEvent{
+					PeerID: connectEvent.PeerID,
+					Type:   connectEvent.Type,
+				})
+			}()
+		},
+	}
+}
+
+// hasPendingStatus checks if a peer is pending handshake [Thread safe]
+func (i *IdentityService) hasPendingStatus(id peer.ID) bool {
+	_, ok := i.pendingPeerConnections.Load(id)
+
+	return ok
+}
+
+// removePendingStatus removes the pending status from a peer,
+// and updates adequate counter information [Thread safe]
+func (i *IdentityService) removePendingStatus(peerID peer.ID) {
+	if value, loaded := i.pendingPeerConnections.LoadAndDelete(peerID); loaded {
+		direction, ok := value.(network.Direction)
+		if !ok {
+			return
+		}
+
+		i.baseServer.UpdatePendingConnCount(-1, direction)
+	}
+}
+
+// addPendingStatus adds the pending status to a peer,
+// and updates adequate counter information [Thread safe]
+func (i *IdentityService) addPendingStatus(id peer.ID, direction network.Direction) {
+	if _, loaded := i.pendingPeerConnections.LoadOrStore(id, direction); !loaded {
+		i.baseServer.UpdatePendingConnCount(1, direction)
+	}
+}
+
+// disconnectFromPeer disconnects from the specified peer
+func (i *IdentityService) disconnectFromPeer(peerID peer.ID, reason string) {
+	i.baseServer.DisconnectFromPeer(peerID, reason)
+}
+
+// handleConnected handles new network connections (handshakes)
+func (i *IdentityService) handleConnected(peerID peer.ID, direction network.Direction) error {
+	clt, clientErr := i.baseServer.NewIdentityClient(peerID)
+	if clientErr != nil {
+		return fmt.Errorf(
+			"unable to create new identity client connection, %w",
+			clientErr,
+		)
+	}
+
+	status := i.constructStatus(peerID)
+	status.Metadata[PeerID] = i.hostID.Pretty()
```

Comment:
This can go inside the body of `constructStatus`.

Reply:
Fixed 👍 

[f763b11](https://github.com/0xPolygon/polygon-edge/pull/343/commits/f763b1182ebfbe0896f36db25fd8be46b5df4e26)